internal/services/voiceservices/voice_services_communications_gateway_resource.go

```diff
@@ -0,0 +1,616 @@
+		output = append(output, platform)
+	}
+
+	return output
+}
+
+func flattenServiceRegionPropertiesModel(inputList *[]communicationsgateways.ServiceRegionProperties) []ServiceRegionPropertiesModel {
+	outputList := make([]ServiceRegionPropertiesModel, 0)
+	if inputList == nil {
+		return outputList
+	}
+
+	for _, input := range *inputList {
+		output := ServiceRegionPropertiesModel{
+			Name: input.Name,
+		}
+
+		output.PrimaryRegionProperties = flattenPrimaryRegionPropertiesModel(&input.PrimaryRegionProperties)
+
+		outputList = append(outputList, output)
+	}
+
+	return outputList
+}
+
+func flattenPrimaryRegionPropertiesModel(input *communicationsgateways.PrimaryRegionProperties) []PrimaryRegionPropertiesModel {
+	outputList := make([]PrimaryRegionPropertiesModel, 0)
+	if input == nil {
+		return outputList
+	}
+
+	output := PrimaryRegionPropertiesModel{
+		OperatorAddresses: input.OperatorAddresses,
+	}
+
+	if input.AllowedMediaSourceAddressPrefixes != nil {
+		output.AllowedMediaSourceAddressPrefixes = *input.AllowedMediaSourceAddressPrefixes
+	}
+
+	if input.AllowedSignalingSourceAddressPrefixes != nil {
+		output.AllowedSignalingSourceAddressPrefixes = *input.AllowedSignalingSourceAddressPrefixes
+	}
+
+	if input.EsrpAddresses != nil {
+		output.EsrpAddresses = *input.EsrpAddresses
+	}
+
+	return append(outputList, output)
+}
+
+func flattenCommunicationsPlatformModel(input []communicationsgateways.CommunicationsPlatform) []string {
+	output := make([]string, 0)
+	if len(input) == 0 {
+		return nil
+	}
+
+	for _, v := range input {
+		output = append(output, string(v))
+	}
+
+	return output
+}
+
+func serviceLocations() *schema.Schema {
+	return &schema.Schema{
+		Type:     pluginsdk.TypeSet,
+		Required: true,
+		Elem: &pluginsdk.Resource{
+			Schema: map[string]*pluginsdk.Schema{
+				"name": {
+					Type:         pluginsdk.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
+				},
+
+				"primary_region_properties": {
+					Type:     pluginsdk.TypeSet,
+					Required: true,
+					MaxItems: 1,
+					Elem: &pluginsdk.Resource{
+						Schema: map[string]*pluginsdk.Schema{
+							"operator_addresses": {
+								Type:     pluginsdk.TypeSet,
+								Required: true,
+								Elem: &pluginsdk.Schema{
+									Type: pluginsdk.TypeString,
+								},
+							},
+							"allowed_media_source_address_prefixes": {
+								Type:     pluginsdk.TypeSet,
+								Optional: true,
+								Elem: &pluginsdk.Schema{
+									Type: pluginsdk.TypeString,
+								},
+							},
+
+							"allowed_signaling_source_address_prefixes": {
+								Type:     pluginsdk.TypeSet,
+								Optional: true,
+								Elem: &pluginsdk.Schema{
+									Type: pluginsdk.TypeString,
+								},
+							},
+
+							"esrp_addresses": {
+								Type:     pluginsdk.TypeSet,
+								Optional: true,
+								Elem: &pluginsdk.Schema{
+									Type: pluginsdk.TypeString,
+								},
```

Comment:
These should be flattened and brought a level up, so under `service_locations`

Reply:
Thanks, the code has been updated. Could you please take another look?